specification/under_development/profiles/Adaptive_Profile.md

```diff
@@ -0,0 +1,963 @@
+Attribute groups are numbered in such a way that they can be easily referenced
+by multiple function definitions or global variables. Consumers of adaptive
+profile-compliant programs must not rely on a particular numbering but instead,
+look for the function to which an attribute with the name `"entry_point"` is
+attached to determine which one to invoke when the program is launched.
+
+Both the `"entry_point"` attribute and the `"output_labeling_schema"` attribute
+can only be attached to a function definition; they are invalid on a function
+that is declared but not defined. For the adaptive profile, this implies that they
+can occur only in one place.
+
+Within the restrictions imposed by the adaptive profile, the number of qubits that
+are needed to execute the quantum program must be known at compile time. This
+number is captured in the form of the `"required_num_qubits"` attribute attached
+to the entry point. The value of the attribute must be the string representation
+of a non-negative 64-bit integer constant.
+
+Similarly, the number of measurement results that need to be stored when
+executing the entry point function is captured by the `"required_num_results"`
+attribute.
+
+Beyond the entry point-specific requirements related to attributes, custom
+attributes may optionally be attached to any of the declared functions. The
+`irreversible` attribute in particular impacts how the program logic in the
+entry point is structured, as described in the section about the [entry point
+definition](#entry-point-definition). Furthermore, the following [LLVM
+attributes](https://llvm.org/docs/LangRef.html#function-attributes) may be used
+according to their intended purpose on function declarations and call sites:
+`inlinehint`, `nofree`, `norecurse`, `readnone`, `readonly`, `writeonly`, and
+`argmemonly`.
+
+## Module Flags Metadata
+
+The following [module
+flags](https://llvm.org/docs/LangRef.html#module-flags-metadata) must be added
+to the QIR bitcode:
+
+- a flag with the string identifier `"qir_major_version"` that contains a
+  constant value of type `i32`
+- a flag with the string identifier `"qir_minor_version"` that contains a
+  constant value of type `i32`
+- a flag with the string identifier `"dynamic_qubit_management"` that contains a
+  constant `true` or `false` value of type `i1`
+- a flag with the string identifier `"dynamic_result_management"` that contains
+  a constant `true` or `false` value of type `i1`
+- A flag named `"qubit_resetting"` with a boolean i1 value indicating
+if the program uses reset operations on qubits.
+- A flag named `"classical_ints"`  with a boolean i1 value indicating
+if the program uses classical computations on integers.
+- A flag named `"classical_floats"`  with a boolean i1 value indicating
+if the program uses classical computations on floating point values.
+- A flag named `"classical_fixed_points"`  with a boolean i1 value indicating
```

Comment:
@peter-campora @qci-amos for specifying fixed points, I think we need a pair of ints, one for the overall bitwidth and another for scaling factor. Or, do we allow varying bit widths with only a single scaling factor for an entire program?